### PR TITLE
CR : Read Permission Handling

### PR DESCRIPTION
**Pull request by RawatH**
PermissionActivity added to check and handle the read permission and pass the permission status back to CommCare App.

### CODE REVIEW

**Comment by shubham1g5** [comment]
Hey @RawatH can we add screenshots/video in PR description for any new UI added as part of this PR ?

File: app/src/main/java/org/commcare/dalvik/reminders/MainActivity.kt [hunk 1]
```diff
@@ -1,44 +1,64 @@
...
 
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
         setContentView(R.layout.activity_main)
+
+        requestPermissionBtn = findViewById(R.id.requestSetting)
+        requestPermissionBtn.setOnClickListener {
+            launchPermissionActivity()
+        }
+
```
Comment (shubham1g5) [inline app/src/main/java/org/commcare/dalvik/reminders/MainActivity.kt:59]: We should do a default permission request on app startup without needing the user to click a button. The button should only be visible if the permission is not granted after the default check.

File: app/src/main/java/org/commcare/dalvik/reminders/MainActivity.kt [hunk 2]
```diff
@@ -50,7 +70,11 @@ class MainActivity : AppCompatActivity() {
                     intent.action = Settings.ACTION_REQUEST_SCHEDULE_EXACT_ALARM
                     startActivityForResult(intent, ALARM_PERMISSION_REQUEST_CODE)
                 }
+            }else{
+                validateReadPermissionsAndSync()
             }
+        }else{
+            validateReadPermissionsAndSync()
         }
```
Comment (shubham1g5) [inline app/src/main/java/org/commcare/dalvik/reminders/MainActivity.kt]: nit: formatting

**Review by shubham1g5 — CHANGES_REQUESTED** [review]

File: app/src/main/java/org/commcare/dalvik/reminders/MainActivity.kt (hunk 1 above)
Comment (RawatH) [reply app/src/main/java/org/commcare/dalvik/reminders/MainActivity.kt:59]: The behaviour is exactly the same.

As the application starts up it checks for alarm premission i.e **validateAlarmPermission()** and once the alarm permission is granted or not required then read permission is intiated automatically.

The "Request Permission" button is only available when the read permission is not granted by user.

File: app/src/main/java/org/commcare/dalvik/reminders/MainActivity.kt (hunk 2 above)
Comment (RawatH) [reply app/src/main/java/org/commcare/dalvik/reminders/MainActivity.kt]: Done

File: app/src/main/java/org/commcare/dalvik/reminders/PermissionActivity.kt [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+        grantPermissionBtn = findViewById(R.id.allowReadPermission)
+
+        grantPermissionBtn.setOnClickListener {
+            requestMultiplePermissionLauncher.launch(CC_CASE_READ_PERMISSION)
+        }
+
+        val backBtn = findViewById<Button>(R.id.backBtn)
+        backBtn.setOnClickListener {
+            dispatchResult(PermissionUtil.hasReadPermission(this))
+        }
```
Comment (shubham1g5) [inline app/src/main/java/org/commcare/dalvik/reminders/PermissionActivity.kt:70]: Users can press back on phone screen/hardware without explicitly clicking on the back button on reminders app UI. Should we in that case do this instead on `onbackPressed` ?

File: app/src/main/java/org/commcare/dalvik/reminders/PermissionActivity.kt [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+    }
+
+    private fun updateUI(permissionMsg: String, btnVisibility: Int = View.GONE) {
+        permissionMsgView.text = permissionMsg
+        grantPermissionBtn.visibility = btnVisibility
+    }
+
+    private fun dispatchResult(isGranted: Boolean) {
+        val resultIntent = Intent().apply {
+            putExtra("permissionName", CC_CASE_READ_PERMISSION)
```
Comment (shubham1g5) [inline app/src/main/java/org/commcare/dalvik/reminders/PermissionActivity.kt:95]: `permissionName` doesn't seem to be getting used from the intent result.

File: app/src/main/java/org/commcare/dalvik/reminders/PermissionActivity.kt (hunk 3 above)
Comment (shubham1g5) [reply app/src/main/java/org/commcare/dalvik/reminders/PermissionActivity.kt:70]: or why do we even have the explicit close button when user can press the back button to get out of permission UI ?

**Review by shubham1g5 — APPROVED** [review]
Looks good to merge but would be nice to get clarifications on the "Close" button UI.

**Comment by RawatH** [comment]
> Looks good to merge but would be nice to get clarifications on the "Close" button UI.


> Looks good to merge but would be nice to get clarifications on the "Close" button UI.

"Close" button was added just for a clear visibility for user to move out of permission activity.

**Comment by shubham1g5** [comment]
> "Close" button was added just for a clear visibility for user to move out of permission activity.

Don't we need to disable the back button press in that case ? Or do activity result handling on pressing `back` as well

File: app/src/main/java/org/commcare/dalvik/reminders/MainActivity.kt [hunk 5]
```diff
@@ -1,44 +1,64 @@
...
 
     private val ALARM_PERMISSION_REQUEST_CODE: Int = 1111;
     private lateinit var reminderViewModel: ReminderViewModel
+    private lateinit var requestPermissionBtn: Button
+
+    private val startPermissionActivityForResult =
+        registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result ->
+            if (result.resultCode == Activity.RESULT_OK) {
+                result.data?.let { intent ->
+                    intent.getStringExtra("isPermissionGranted")?.let {permissionResult ->
```
Comment (shubham1g5) [inline app/src/main/java/org/commcare/dalvik/reminders/MainActivity.kt:39]: `isPermissionGranted` needs to be returned back to the CC when user closes Reminders App as well.

**Review by shubham1g5 — CHANGES_REQUESTED** [review]

File: app/src/main/AndroidManifest.xml [hunk 6]
```diff
@@ -21,6 +21,16 @@
             </intent-filter>
         </activity>
 
+
+        <activity android:name=".PermissionActivity"
+            android:exported="true">
+            <intent-filter>
+                <action android:name="custom.intent.permission.request" />
```
Comment (shubham1g5) [inline app/src/main/AndroidManifest.xml]: we should rename the intent action here to - `org.commcare.dalvik.reminders.permission`

File: app/src/main/java/org/commcare/dalvik/reminders/MainActivity.kt (hunk 5 above)
Comment (shubham1g5) [reply app/src/main/java/org/commcare/dalvik/reminders/MainActivity.kt:39]: I see that we are exposing the PermissionActivity directly instead.

File: app/src/main/AndroidManifest.xml (hunk 6 above)
Comment (RawatH) [reply app/src/main/AndroidManifest.xml]: Done

**Comment by RawatH** [comment]
> > "Close" button was added just for a clear visibility for user to move out of permission activity.
> 
> Don't we need to disable the back button press in that case ? Or do activity result handling on pressing `back` as well

Device and Close button have some functionality. They will check the permission status and dispatch the result.

**Review by shubham1g5 — APPROVED** [review]
Thanks! Looks great.

**Pull request merged**